src/trio/_core/_asyncgens.py

```diff
@@ -78,12 +80,14 @@ def finalize_in_trio_context(
                 self.trailing_needs_finalize.add(agen)
 
         def finalizer(agen: AsyncGeneratorType[object, NoReturn]) -> None:
```

Comment:
I think this can only work when using https://github.com/python-trio/trio/pull/3110 because if self.foreign.remove(id(agen)) is interrupted with a KI the id will outlive the agen and could end up being re-used